include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+  }
+
+  /**
+   * @brief Mark all blocks that overlap the given range as unavailable
+   */
+  void mark_as_unavailable(void *ptr, size_t size) {
+    char *cptr = static_cast<char *>(ptr);
+    char *start = detail::align_ptr_down(cptr, block_size_);
+    char *end = detail::align_ptr(cptr + size, block_size_);
+    set_block_avaialbility(start, end, false);
+  }
+
+
+  void set_block_avaialbility(char *start, char *end, bool availability) {
+    if (end <= start)  // NOTE it's possible that end < start - see mark_as_available
+      return;
+    va_region *va = va_find(start);
+    assert(va != nullptr);
+    assert(va == va_find(end-1));
+    va->set_block_availability(start, end, availability);
+  }
+
+  void map_storage(void *ptr, size_t size) {
+    CUdeviceptr dptr = reinterpret_cast<CUdeviceptr>(ptr);
+    auto *region = va_find(ptr);
+    assert(region);
+    CUdeviceptr region_start  = region->address_range.ptr();
+    CUdeviceptr region_end    = region->address_range.end();
+    ptrdiff_t offset = dptr - region_start;
+    int start_block_idx = offset / block_size_;
+    int end_block_idx = (offset + size + block_size_ - 1) / block_size_;
+    struct block_range {
+      int begin, end;
+    };
+    SmallVector<block_range, 8> to_map;
+    int blocks_to_map = 0;
+    for (int block_idx = start_block_idx; block_idx < end_block_idx; ) {
+      int next_unmapped = region->mapped.find(false, block_idx);
+      if (next_unmapped >= end_block_idx)
+        break;  // everything we need is mapped)
+      block_idx = next_unmapped;
+      int next_mapped = region->mapped.find(true, next_unmapped + 1);
+      int next = std::min(end_block_idx, next_mapped);
+      blocks_to_map += next - next_unmapped;
+      block_idx = next;
+      to_map.push_back({ next_unmapped, block_idx });
+    }
+
+    mark_as_unavailable(ptr, size);
+
+    SmallVector<cuvm::CUMem, 256> free_blocks;
+    get_free_blocks(free_blocks, blocks_to_map);
+    assert(static_cast<int>(free_blocks.size()) == blocks_to_map);
+
+    int i = 0;
+    for (const block_range &br : to_map) {
+      region->available.fill(br.begin, br.end, true);
+      for (int b = br.begin; b < br.end; b++) {
+        assert(i < static_cast<int>(free_blocks.size()));
+        region->map_block(b, std::move(free_blocks[i++]));
+      }
+      size_t range_size = (br.end - br.begin) * block_size_;
+      free_mapped_.put(region->block_ptr(br.begin), range_size);
+      stat_add_free(range_size);
+    }
+    mark_as_available(ptr, size);
+    assert(i == blocks_to_map);
+  }
+
+  /**
+   * @brief Obtains `count` physical storage blocks.
+   *
+   * Obtains `count` physical storage blocks and places them in the collection `out`.
+   * The blocks might be obtained either by:
+   * - unmapping existing mapped but free blocks, if avaialble
+   * - allocating new blocks.
+   */
+  template <typename Collection>
+  void get_free_blocks(Collection &out, int count) {
```

Comment:
```suggestion
  void get_free_blocks(Collection &out, size_t count) {
```

Reply:
Block count is always represented by `int` here whereas `size_t` is used for memory sizes (in bytes). Given that the blocks are at least 2MiB (that's the granularity) I can hardly imagine exceeding `int` (that would be 2EiB of memory).

Reply:
Does it make sense to make it unsigned?